### PR TITLE
[GEOT-7180] GeoJSONReader ignores feature "id" value.

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOT-7180](https://osgeo-org.atlassian.net/browse/GEOT-7180) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

cleanup

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/unsupported/geojson-core/src/test/java/org/geotools/data/geojson/GeoJSONReaderTest.java [hunk 1]
```diff
@@ -177,7 +177,7 @@ public void testFeatureCollectionParser() throws Exception {
         f = DataUtilities.first(features);
 
         assertNotNull(f);
-        assertEquals("features.0", f.getID());
+        assertEquals("feature.0", f.getID());
```
Comment (ianturton) [inline modules/unsupported/geojson-core/src/test/java/org/geotools/data/geojson/GeoJSONReaderTest.java:180]: I don't like that these have changed, is there a reason for the loss of the `s`?

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java [hunk 2]
```diff
@@ -436,6 +451,24 @@ private SimpleFeature getNextFeature(ObjectNode node) throws IOException {
         return feature;
     }
 
+    private String getId(ObjectNode node) {
+        String newId = null;
+        if (node.has(ID)) {
+            if (idPrefix != null) {
+                newId = idPrefix + "." + node.get(ID).asText();
+            } else {
+                newId = node.get(ID).asText();
```
Comment (ianturton) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:460]: I think this should still have the `basename` prefixing the `ID` attribute value

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java [hunk 3]
```diff
@@ -118,6 +122,18 @@ public GeoJSONReader(URL url) throws IOException {
...
+    /**
+     * Builds a GeoJSON parser from a GeoJSON source, located at the specified URL.
+     *
+     * @param url String representation of the URL
+     * @param idPrefix prefix to use for feature ids
+     * @throws IOException if the URL is not valid
+     */
+    public GeoJSONReader(URL url, String idPrefix) throws IOException {
+        this(url);
+        this.idPrefix = idPrefix;
```
Comment (ianturton) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:134]: don't you need to call `this(url)` here or initialize the `parser`?

File: modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java [hunk 4]
```diff
@@ -61,8 +62,8 @@ public GeoJSONDataStore(File f) {
     }
 
     GeoJSONReader read() throws IOException {
-        GeoJSONReader reader = new GeoJSONReader(getUrl());
-
+        GeoJSONReader reader =
+                new GeoJSONReader(getUrl(), FilenameUtils.getBaseName(url.getPath()));
```
Comment (ianturton) [inline modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: Why is this necessary?

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 2 above)
Comment (ianturton) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:460]: What happens if the `ID` attributes are not unique?

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java [hunk 5]
```diff
@@ -86,6 +86,8 @@ public class GeoJSONReader implements AutoCloseable {
 
     private SimpleFeatureBuilder builder;
 
+    private static final String ID = "id";
```
Comment (ianturton) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:89]: The user should be able to choose this value, it could be `fid` or `gid` etc

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java [hunk 6]
```diff
@@ -411,8 +427,7 @@ private SimpleFeature getNextFeature(ObjectNode node) throws IOException {
                 }
             }
             builder.set(GEOMETRY_NAME, g);
-
-            String newId = baseName + "." + nextID++;
+            String newId = getId(node);
```
Comment (ianturton) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:430]: If we are using the `ID` attribute then it should be excluded from the feature that is created.

**Review by ianturton — CHANGES_REQUESTED** [review]

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 2 above)
Comment (aaime) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:460]: I'd argue a low-level GeoJSONReader should be truthful to the GeoJSON content, and be allowed to return exactly the id found in the document. Having an option to use the prefix, for datastore usage, allows for compatibilty.

File: modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java (hunk 4 above)
Comment (aaime) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: To make it prefix the ids with the file name, and retain the same behavior.

Comment (ianturton) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: but that is what the `new GeoJSONReader(getUrl());` call did any way - so I don't see any need for this

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 2 above)
Comment (aaime) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:460]: My first reaction would have been "then the GeoJSON is broken". But it turns out the specification even allows the id to be missing and does not talk about uniqueness.

However this can be restricted by specifications that build on top of GeoJSON. For example, the [STAC-item specification](https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md#item-fields) mandates "The ID should be unique within the [Collection](https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md) that contains the Item."

So it seems we likely need some id generation strategies that a caller can use:

- Existing one, file name + progressive, for random GeoJSON files that use the full liberty of the spec.
- Using exactly the id found in the features, for contexts where uniqueness is guaranteed
- Also keep the file name + id, to make the ids "more globally unique" (needed for the STAC datastore, to make it behave like other GeoServer stores).

Comment (ianturton) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:460]: OK, that makes sense

File: modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java (hunk 4 above)
Comment (aaime) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: Ah, did not notice, sorry.

Comment (turingtestfail) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: Yeah, I added the second argument as a means to seperate the featurebuilder name from the prefix.  If those should always be the same I can remove the extra constructor argument.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 5 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:89]: The standard is pretty specific about this:
If a Feature has a commonly used identifier, that identifier
      SHOULD be included as a member of the Feature object with the name
      "id", and the value of this member is either a JSON string or
      number.

File: modules/unsupported/geojson-core/src/test/java/org/geotools/data/geojson/GeoJSONReaderTest.java (hunk 1 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/test/java/org/geotools/data/geojson/GeoJSONReaderTest.java]: Originally changed to match GeoJSON in the tests.  I have gone ahead and added the s to JSON and reverted the asserts.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 3 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:134]: See comment below about the featurebuilder name and the prefix.  If a seperate prefix is not needed, this new constructor will also go away.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 6 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java]: Yup, ambiguous method name here.  Changed to reflect that it either gets the ID or generates one if not found.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 5 above)
Comment (aaime) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java:89]: Yes, but SHOULD is not mandatory, see the RFC titled "[Key words for use in RFCs to Indicate Requirement Levels](https://datatracker.ietf.org/doc/html/rfc2119)", quoting:

```
SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.
```

So strong recommendation, but not an absolute requirement. Also, the wording talks about identifier, but does not say anything about its uniqueness and in which context (e.g., in the STAC spec the id must be unique only inside that collection). To make another example, I identify as Andrea, but it's too common, Andrea Aime is better, it's (accidentaly) unique in the province where I live, but if I want uniqueness at the country level, I have to switch to the SSN.

File: modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java (hunk 4 above)
Comment (ianturton) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: The idea of a 2nd argument is fine, it just isn't needed here. as if you don't supply the 2nd argument the other constructor will make a baseName call.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 6 above)
Comment (ianturton) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java]: I was saying that if we use the `id` attribute as the feature ID then we should not return it as an attribute too

File: modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java (hunk 4 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: Ahh, gotcha.  It is now needed because without the second argument, if there is an existing id field, no prefix is added. In cases where there is no existing id field, it behaves as before and defaults to adding the derived prefix.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 5 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java]: Gotcha.  Updated to add constructor that allows specification of an id field name.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 6 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java]: Yup, my mistake.  Fixed to exclude id field from general attribute map.

File: modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java (hunk 4 above)
Comment (ianturton) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java:66]: As I said, I understand why you have provided a constructor with 2 arguments - my question is why in this exact line is it necessary to pass in the basename as the 2nd argument when the single argument constructor will do exactly the same thing?

**Comment by aaime** [comment]
@turingtestfail there is a small conflict.

File: modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java (hunk 4 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-store/src/main/java/org/geotools/data/geojson/store/GeoJSONDataStore.java]: Now changed so the default setting mean that this is no longer needed, as requested.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java (hunk 2 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/data/geojson/GeoJSONReader.java]: Implemented, as described.

**Comment by aaime** [comment]
Seems reasonable now, @ianturton what do you think?

**Review by ianturton — APPROVED** [review]
